### ISSUE TITLE
btrbk fails to recognize that it can incrementally send in the reverse direction

### ISSUE THREAD

**Issue by ian-kelling** [digint/btrbk#114]
Btrfs allows this, which is very useful for the case where the backup system is a failover, or for whatever reason becomes primary, and the old primary system becomes the backup . Here is an example, omiting the the "btrfs" part of the commands:

```
mkdir -p a b
sub create a/rw
# modify a/rw
sub snap -r a/rw a/1
send a/1 | receive b
# modify a/rw
sub snap -r a/rw a/2
send -p a/1 a/2 | receive b
# imagine a goes offline, b becomes primary
sub snap b/2 b/rw
# modify b/rw
sub snap -r b/rw b/3
# imagine a comes back online, we do an incremental send back:
send -p b/2 b/3 | receive a
```
btrbk, instead of doing the incremental last send, instead will do a full send: "send b/3 | receive a".
It can be detected from parent/receive metadata that it can do an incremental send.

Here is a debug run when it should do an incremental send back, canceled after the send starts.
I manually ran the incremental send after deleting the partially sent subvolume and it worked fine.
```
# btrbk -l debug -r run
btrbk command line client, version 0.23.3  (Fri Nov 18 20:08:47 2016)
Using configuration: /etc/btrbk.conf
Initializing volume section: /mnt/root
### 'btrfs' 'subvolume' 'show' '/mnt/root'
Command execution successful
found btrfs root: /mnt/root
### 'btrfs' 'subvolume' 'list' '-a' '-c' '-u' '-q' '-R' '/mnt/root'
Command execution successful
Parsed 22 total subvolumes for filesystem at: /mnt/root
### 'btrfs' 'subvolume' 'list' '-a' '-r' '/mnt/root'
Command execution successful
Parsed 18 readonly subvolumes for filesystem at: /mnt/root
Initializing subvolume section: /mnt/root/q
Found "/mnt/root/q" (id=644) in btrfs subtree of: /mnt/root
Initializing target section: treetowl:/mnt/root/btrbk
### 'ssh' '-i' '/root/.ssh/id_rsa' 'root@treetowl' 'btrfs' 'subvolume' 'show' '/mnt/root/btrbk'
Command execution failed (exitcode=1): "'ssh' '-i' '/root/.ssh/id_rsa' 'root@treetowl' 'btrfs' 'subvolume' 'show' '/mnt/root/btrbk' 2>&1"
Resolving btrfs mount point for: treetowl:/mnt/root/btrbk
### 'ssh' '-i' '/root/.ssh/id_rsa' 'root@treetowl' 'readlink' '-e' '/mnt/root/btrbk'
Command execution successful
Real path for "treetowl:/mnt/root/btrbk" is: /mnt/root/btrbk
### 'ssh' '-i' '/root/.ssh/id_rsa' 'root@treetowl' 'cat' '/proc/self/mounts'
Command execution successful
Found btrfs mount point for "treetowl:/mnt/root/btrbk": /mnt/root (subvolid=5)
### 'ssh' '-i' '/root/.ssh/id_rsa' 'root@treetowl' 'btrfs' 'subvolume' 'list' '-a' '-c' '-u' '-q' '-R' '/mnt/root'
Command execution successful
Parsed 36 total subvolumes for filesystem at: treetowl:/mnt/root
### 'ssh' '-i' '/root/.ssh/id_rsa' 'root@treetowl' 'btrfs' 'subvolume' 'list' '-a' '-r' '/mnt/root'
Command execution successful
Parsed 29 readonly subvolumes for filesystem at: treetowl:/mnt/root
Opening syslog
Syslog enabled
Skipping snapshot creation (option "-r" present)
Found 1 snapshot children of "/mnt/root/q" in: /mnt/root/btrbk/q.*
Checking for missing backups of subvolume "/mnt/root/q" in "treetowl:/mnt/root/btrbk/"
Adding backup candidate: /mnt/root/btrbk/q.20161118T193925-0800
Checking schedule for backup candidates
Schedule: preserving all within 6 hours; first of day for 14 days; first daily in week (starting on sunday) for 8 weeks; first weekly of month for 24 months
Scheduler reference time: 2016-11-18T20:08:47-0800
Preserving 1/1 items
Creating subvolume backup (send-receive) for: /mnt/root/btrbk/q.20161118T193925-0800
No common subvolumes of "/mnt/root/btrbk/q.20161118T193925-0800" found in src="/mnt/root/", target="treetowl:/mnt/root/btrbk/"
No common parent subvolume present, creating full backup...
[send/receive] source: /mnt/root/btrbk/q.20161118T193925-0800
[send/receive] target: treetowl:/mnt/root/btrbk/q.20161118T193925-0800
### 'btrfs' 'send' '/mnt/root/btrbk/q.20161118T193925-0800' | 'ssh' '-i' '/root/.ssh/id_rsa' 'root@treetowl' 'btrfs' 'receive' '/mnt/root/btrbk/'
^C
ERROR: Cought SIGINT, dumping transaction log:
localtime type status duration target_url source_url parent_url message
2016-11-18T20:08:49-0800 startup v0.23.3 - - - - # btrbk command line client, version 0.23.3
2016-11-18T20:08:49-0800 send-receive starting - ssh://treetowl/mnt/root/btrbk/q.20161118T193925-0800 /mnt/root/btrbk/q.20161118T193925-0800 - -
2016-11-18T20:08:55-0800 signal SIGINT - - - - -
```
The configs used in this case for each host were the same, just with different hostname for target send-receive.

**Comment by ian-kelling** [comment]
@digint, let me know if you are going to work on fixing this, or else I'm gonna take a crack at fixing it myself. I'm sure you can do it faster :)

**Comment by digint** [comment]
For now, the only candidates for a incremental send of a rw source are the "direct children of the rw source subvolume" ([L2296](https://github.com/digint/btrbk/blob/master/btrbk#L2296)), which is "b/3" in your example. For some reason, the direct parent of a rw source is not taken as candidate, this is explicitely done only for ro sources ([L2271](https://github.com/digint/btrbk/blob/master/btrbk#L2271), [L2268](https://github.com/digint/btrbk/blob/master/btrbk#L2268)).
There is probably a good reason for this, I can't recall exactly why right now (or maybe it's just legacy or paranoia, and I thought that it would not make sense to check the parents or rw sources).

I order to be ultra robust, btrbk should even recursively walk all parents and add them to the candidate list?

I will investigate this more deeply some day next week (pretty busy at the moment...), let me know if you have some findings in the meantime.

**Comment by ian-kelling** [comment]
Thanks for the hints. I am gonna poke around in the perl.

**Comment by digint** [comment]
fixed in: d02f67a9248c6eac5e918ab4d43056baf2d5c11e

Thanks @ian-kelling for contributing!
